pex/specifier_sets.py

```diff
@@ -0,0 +1,331 @@
+        elif "~=" == spec.operator:
+            yield LowerBound(version, inclusive=True)
+            # The SpecifierSet argument type ensures we have at least X.Y across all vendored
+            # versions of packaging we ship; so the -2 index access is safe.
+            ceiling = _increment_release(version, -2)
+            yield UpperBound(ceiling, inclusive=False)
+
+
+def as_range(specifier_set):
+    # type: (Union[str, SpecifierSet]) -> Union[ArbitraryEquality, Range]
+
+    lower_bounds = []  # type: List[LowerBound]
+    upper_bounds = []  # type: List[UpperBound]
+    excludes = []  # type: List[ExcludedRange]
+    for bound in _bounds(
+        specifier_set if isinstance(specifier_set, SpecifierSet) else SpecifierSet(specifier_set)
+    ):
+        if isinstance(bound, ArbitraryEquality):
+            return bound
+        elif isinstance(bound, ExcludedRange):
+            excludes.append(bound)
+        elif isinstance(bound, LowerBound):
+            lower_bounds.append(bound)
+        else:
+            upper_bounds.append(bound)
+
+    # Narrow bounds to the smallest range.
+    lower = sorted(lower_bounds)[-1] if lower_bounds else None
+    upper = sorted(upper_bounds)[0] if upper_bounds else None
+
+    # Further narrow bounds by any overlapping excludes.
+    if lower:
+        for exclude in sorted(excludes, key=lambda ex: ex.lower):
+            # MyPy fails to typecheck <= under Python 2.7 only, even though _LowerBound (via its
+            # _Bound base class) has @total_ordering applied.
+            if exclude.lower <= lower:  # type: ignore[operator]
+                new_lower = LowerBound(exclude.upper.version, inclusive=not exclude.upper.inclusive)
+                if new_lower > lower:
+                    lower = new_lower
+                    excludes.remove(exclude)
+    if upper:
+        for exclude in sorted(excludes, key=lambda ex: ex.upper, reverse=True):
+            # MyPy fails to typecheck >= under Python 2.7 only, even though _UpperBound (via its
+            # _Bound base class) has @total_ordering applied.
+            if exclude.upper >= upper:  # type: ignore[operator]
+                new_upper = UpperBound(exclude.lower.version, inclusive=not exclude.lower.inclusive)
+                if new_upper < upper:
+                    upper = new_upper
+                    excludes.remove(exclude)
+
+    # Only retain excludes whose range resides inside the final fully narrowed bounds.
+    for exclude in tuple(excludes):
+        if upper:
+            if exclude.lower.version > upper.version or (
+                exclude.lower.version == upper.version
+                and exclude.lower.inclusive
+                and upper.inclusive
+            ):
+                excludes.remove(exclude)
+                continue
+        if lower:
+            if exclude.upper.version < lower.version or (
+                exclude.upper.version == lower.version and lower.inclusive
```

Comment:
why is not `exclude.upper.inclusive` consulted? (or should it be `and not lower.inclusive`?)

Reply:
Fixed by eliminating this code altogether.